infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java

```diff
@@ -0,0 +1,210 @@
+package org.corfudb.infrastructure.log.statetransfer.batchprocessor.protocolbatchprocessor;
+
+import com.google.common.collect.Ordering;
+import com.google.common.collect.Sets;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.common.result.Result;
+import org.corfudb.infrastructure.log.StreamLog;
+import org.corfudb.infrastructure.log.statetransfer.StateTransferException;
+import org.corfudb.infrastructure.log.statetransfer.batch.Batch;
+import org.corfudb.infrastructure.log.statetransfer.batch.BatchResult;
+import org.corfudb.infrastructure.log.statetransfer.batch.ReadBatch;
+import org.corfudb.infrastructure.log.statetransfer.batchprocessor.StateTransferBatchProcessor;
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.protocols.wireprotocol.LogData;
+import org.corfudb.runtime.exceptions.RetryExhaustedException;
+import org.corfudb.runtime.view.AddressSpaceView;
+import org.corfudb.runtime.view.ReadOptions;
+import org.corfudb.util.retry.ExponentialBackoffRetry;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.RetryNeededException;
+
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static lombok.Builder.Default;
+import static org.corfudb.infrastructure.log.statetransfer.batch.BatchResult.FailureStatus.FAILED;
+
+/**
+ * A batch processor that transfers non-committed addresses one batch at a time
+ * via a replication protocol.
+ */
+@Slf4j
+@Builder
+public class ProtocolBatchProcessor implements StateTransferBatchProcessor {
+
+    /**
+     * Configurations for the retry logic.
+     */
+    @Default
+    private final int maxRetries = 3;
+    @Default
+    private final Duration maxRetryTimeout = Duration.ofSeconds(10);
+    @Default
+    private final float randomFactorBackoff = 0.5f;
+
+    /**
+     * Default read options for the replication protocol read.
+     */
+    @Getter
+    private static ReadOptions readOptions = ReadOptions.builder()
+            .waitForHole(true)
+            .clientCacheable(false)
+            .serverCacheable(false)
+            .build();
+
+    @Getter
+    public final StreamLog streamLog;
+
+    @Getter
+    private final AddressSpaceView addressSpaceView;
+
+    @Override
+    public CompletableFuture<BatchResult> transfer(Batch batch) {
+        return readRecords(batch, 0)
```

Comment:
I think its sufficient for the readRecords to just retry on readRecords. I don't think we need any fancy error handling logic while writing to the local log for the addresses that we know are not there (because we check beforehand).